File: app/src/main/java/com/fieldbook/tracker/database/SearchQueryBuilder.java
package com.fieldbook.tracker.database;

import android.content.SharedPreferences;
import android.database.DatabaseUtils;

import com.fieldbook.tracker.R;
import com.fieldbook.tracker.activities.CollectActivity;
import com.fieldbook.tracker.objects.SearchDialogDataModel;
import com.fieldbook.tracker.objects.TraitObject;
import com.fieldbook.tracker.preferences.GeneralKeys;
import com.fieldbook.tracker.utilities.CategoryJsonUtil;

import org.brapi.v2.model.pheno.BrAPIScaleValidValuesCategories;

import java.util.ArrayList;
import java.util.List;

public class SearchQueryBuilder {
    public static String TICK = "\"";
    private final CollectActivity originActivity;
    private final List<SearchDialogDataModel> dataSet;
    private StringBuilder queryBuilder;

    public SearchQueryBuilder(CollectActivity originActivity, List<SearchDialogDataModel> dataSet) {
        this.originActivity = originActivity;
        this.dataSet = dataSet;
    }


    public String buildSearchQuery(){
        SharedPreferences ep = originActivity.getPreferences();
        queryBuilder = new StringBuilder();

        // Create the sql query based on user selection
        String attributeSelectStatement = "select ObservationUnitProperty.id, ObservationUnitProperty." + TICK + ep.getString(GeneralKeys.UNIQUE_NAME, "") + TICK + ", " + " ObservationUnitProperty." + TICK + ep.getString(GeneralKeys.PRIMARY_NAME, "") + TICK + "," + " ObservationUnitProperty." + TICK + ep.getString(GeneralKeys.SECONDARY_NAME, "") + TICK + " from ObservationUnitProperty where ObservationUnitProperty.id is not null";
        String traitSelectStatement = "select ObservationUnitProperty.id, ObservationUnitProperty." + TICK + ep.getString(GeneralKeys.UNIQUE_NAME, "") + TICK + ", " + " ObservationUnitProperty." + TICK + ep.getString(GeneralKeys.PRIMARY_NAME, "") + TICK + "," + " ObservationUnitProperty." + TICK + ep.getString(GeneralKeys.SECONDARY_NAME, "") + TICK + " from observation_variables, ObservationUnitProperty, observations where observations.observation_unit_id = ObservationUnitProperty." + TICK + ep.getString(GeneralKeys.UNIQUE_NAME, "") + TICK + " and observations.observation_variable_name = observation_variables.observation_variable_name and observations.observation_variable_field_book_format = observation_variables.observation_variable_field_book_format";

        for (int i = 0; i < dataSet.size(); i++) {

            // Builds the SQL query for the 'i'th user selection
            StringBuilder currentSubQuery = new StringBuilder();

            String column = dataSet.get(i).getAttribute();
            int operator = dataSet.get(i).getImageResourceId();
            String searchText = dataSet.get(i).getText();

            TraitObject traitObject = originActivity.getDatabase().getTraitByName(column);

            boolean isAttribute = false; // Checks if 'column' is an attribute or a trait
            boolean isCategorical = false; // Checks if 'column' is a categorical trait

            if (traitObject == null) {
                isAttribute = true;
                currentSubQuery.append(attributeSelectStatement).append(" and ").append("ObservationUnitProperty." + TICK + column + TICK);
            } else {
                currentSubQuery.append(traitSelectStatement).append(" and ").append("observation_variables.observation_variable_name = " + TICK + column + TICK);

                if (traitObject.getFormat().equals("categorical") || traitObject.getFormat().equals("multicat") || traitObject.getFormat().equals("qualitative")) {
                    isCategorical = true;
                    searchText = encodeCategorical(searchText);
                }
            }

            // This is to prevent crashes when the user uses special characters
            String truncatedSearchText = DatabaseUtils.sqlEscapeString(searchText);

            // Represents the SQL condition based on the operator and search text
            String condition = "";

            // 0: Equals to
            if (operator == R.drawable.ic_tb_equal) {
                if (isAttribute) condition = " = " + truncatedSearchText;
                else condition = " and value = " + truncatedSearchText;
            }

            // 1: Not equals to
            else if (operator == R.drawable.ic_tb_not_equal) {
                if (isAttribute) condition = " != " + truncatedSearchText;
                else condition = " and value != " + truncatedSearchText;
            }

            // 2: Contains
            else if (operator == R.drawable.ic_tb_contains) {
                if (isAttribute)
                    condition = " like " + DatabaseUtils.sqlEscapeString("%" + searchText + "%");
                else {
                    if (!isCategorical)
                        condition = " and observations.value like " + DatabaseUtils.sqlEscapeString("%" + searchText + "%");
                    else {
                        String decodedT = CategoryJsonUtil.Companion.decode(searchText).get(0).getValue();
                        condition = " and observations.value like " + DatabaseUtils.sqlEscapeString("%[{\"label\":\"%" + decodedT + "%\",\"value\":\"%" + decodedT + "%\"}]%");
                    }
                }
            }

            // 3: More than
            else if (operator == R.drawable.ic_tb_greater_than) {
                if (isAttribute) condition = " > " + truncatedSearchText;
                else condition = " and CAST(value AS INTEGER) > " + truncatedSearchText;
            }

            // 4: less than
            else if (operator == R.drawable.ic_tb_less_than) {
                if (isAttribute) condition = " < " + truncatedSearchText;
                else condition = " and CAST(value AS INTEGER) < " + truncatedSearchText;
            }

            currentSubQuery.append(condition);

            if (i == 0) queryBuilder.append(currentSubQuery);
            else intersectQuery(currentSubQuery);
        }
        return queryBuilder.toString();
    }

    public void intersectQuery(StringBuilder query)
    {
        queryBuilder.append(" INTERSECT ").append(query);
    }

    public void unionQuery(String query)
    {
        queryBuilder.append(" UNION ").append(query);
    }

    public void exceptQuery(String query)
    {
        queryBuilder.append(" EXCEPT ").append(query);
    }

    public String encodeCategorical(String t) {
        return CategoryJsonUtil.Companion.encode(new ArrayList<BrAPIScaleValidValuesCategories>() {{
            add(new BrAPIScaleValidValuesCategories().label(t).value(t));
        }});
    }

}

File: app/src/main/java/com/fieldbook/tracker/database/SearchQueryBuilder.kt
package com.fieldbook.tracker.database

import android.content.SharedPreferences
import android.database.DatabaseUtils
import android.util.Log
import com.fieldbook.tracker.R
import com.fieldbook.tracker.adapters.AttributeAdapter
import com.fieldbook.tracker.objects.SearchDialogDataModel
import com.fieldbook.tracker.preferences.GeneralKeys
import com.fieldbook.tracker.utilities.CategoryJsonUtil
import com.fieldbook.tracker.utilities.CategoryJsonUtil.Companion.decode
import org.brapi.v2.model.pheno.BrAPIScaleValidValuesCategories

class SearchQueryBuilder(
    private val preferences: SharedPreferences,
    private val database: DataHelper
) {

    companion object {
        const val TAG = "SearchQueryBuilder"
    }

    private val queryBuilder: StringBuilder by lazy {
        StringBuilder()
    }

    private val uniqueName by lazy {
        preferences.getString(
            GeneralKeys.UNIQUE_NAME,
            ""
        )
    }

    private val primaryName by lazy {
        preferences.getString(
            GeneralKeys.PRIMARY_NAME,
            ""
        )
    }

    private val secondaryName by lazy {
        preferences.getString(
            GeneralKeys.SECONDARY_NAME,
            ""
        )
    }

    fun buildSearchQuery(dataSet: MutableList<SearchDialogDataModel>): String {

        queryBuilder.clear()

        // Create the sql query based on user selection
        val attributeSelectStatement =
            """
                SELECT OUP.id, OUP."$uniqueName", OUP."$primaryName", OUP."$secondaryName"
                FROM ObservationUnitProperty AS OUP
                WHERE OUP.id IS NOT NULL
            """.trimIndent()

        val traitSelectStatement =
            """
                SELECT OUP.id, OUP."$uniqueName", OUP."$primaryName", OUP."$secondaryName"
                FROM ObservationUnitProperty AS OUP
                JOIN observations AS O ON O.observation_unit_id = OUP."$uniqueName"
                JOIN observation_variables AS V ON O.observation_variable_db_id = V.internal_id_observation_variable
            """.trimIndent()

        for (i in dataSet.indices) {

            // Builds the SQL query for the 'i'th user selection
            val currentSubQuery = StringBuilder()

            val model: AttributeAdapter.AttributeModel = dataSet[i].attribute
            val column = model.label
            val operator = dataSet[i].imageResourceId
            var searchText = dataSet[i].text

            val traitObject = model.trait?.id?.let { id -> database.getTraitById(id) }

            var isAttribute = false // Checks if 'column' is an attribute or a trait
            var isCategorical = false // Checks if 'column' is a categorical trait

            if (traitObject == null) {
                isAttribute = true
                currentSubQuery.append(attributeSelectStatement).append(" AND ")
                    .append("OUP.\"$column\"")
            } else {
                currentSubQuery.append(traitSelectStatement).append(" AND ")
                    .append("V.observation_variable_name = \"$column\"")

                if (traitObject.format == "categorical" || traitObject.format == "qualitative") {
                    isCategorical = true
                    searchText = encodeCategorical(searchText)
                }
            }

            // This is to prevent crashes when the user uses special characters
            val truncatedSearchText = DatabaseUtils.sqlEscapeString(searchText)

            // Represents the SQL condition based on the operator and search text
            var condition = ""

            // 0: Equals to
            if (operator == R.drawable.ic_tb_equal) {
                condition = if (isAttribute) " = $truncatedSearchText"
                else " AND value = $truncatedSearchText"
            } else if (operator == R.drawable.ic_tb_not_equal) {
                condition = if (isAttribute) " != $truncatedSearchText"
                else " AND value != $truncatedSearchText"
            } else if (operator == R.drawable.ic_tb_contains) {
                if (isAttribute) condition =
                    " LIKE " + DatabaseUtils.sqlEscapeString("%$searchText%")
                else {
                    if (!isCategorical) condition =
                        " AND O.value LIKE " + DatabaseUtils.sqlEscapeString("%$searchText%")
                    else {
                        val decodedT = decode(searchText)[0].value
                        condition =
                            " AND O.value LIKE " + DatabaseUtils.sqlEscapeString("%[{\"label\":\"%$decodedT%\",\"value\":\"%$decodedT%\"}]%")
                    }
                }
            } else if (operator == R.drawable.ic_tb_greater_than) {
                condition = if (isAttribute) " > $truncatedSearchText"
                else " AND CAST(value AS INTEGER) > $truncatedSearchText"
            } else if (operator == R.drawable.ic_tb_less_than) {
                condition = if (isAttribute) " < $truncatedSearchText"
                else " AND CAST(value AS INTEGER) < $truncatedSearchText"
            }

            currentSubQuery.append(condition)

            if (i == 0) queryBuilder.append(currentSubQuery)
            else intersectQuery(currentSubQuery)
        }

        Log.d(TAG, "Query: $queryBuilder")

        return queryBuilder.toString()
    }

    fun intersectQuery(query: StringBuilder?) {
        queryBuilder.append(" INTERSECT ").append(query)
    }

    fun encodeCategorical(t: String?): String {
        return CategoryJsonUtil.Companion.encode(object :
            ArrayList<BrAPIScaleValidValuesCategories?>() {
            init {
                add(BrAPIScaleValidValuesCategories().label(t).value(t))
            }
        } as ArrayList<BrAPIScaleValidValuesCategories>)
    }
}
